lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
even if you check before and after, nothing prevent having a stop occur while in the sleep and the shutdown sequence will be tied to the sleep period. only way to prevent this is to simply use a "stoppable_sleep" which check for stop periodically.

other than that, we cannot control the code that is not under our control. we can try our best to force stop foreign code by using whichever api they provide by if it does not cooperate it will be caught by the (eventual) stuck shutdown logic, otherwise we can submit upstream patches on 3rd party libraries that do not correctly stopping or closing etc. 

the hope is that such scenarios will be rare. the ones that are obvious are the input plugins that are blocked on IO and that the stop method will need to unblock.